{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 1\n",
    "Date: 2024/12/19\n",
    "The purpose of this script is to take the GFF created by gffread with the following command:\n",
    "\n",
    "gffread -E /home/administrator/Documents/Kaas/Venom_ncRNA_project/Usable_data/Genome_files/CroVir_rnd1.all.maker.final.homologIDs.updatedNov2019_with_myos_geneidmod_edited_with_BPP.gtf -o /home/administrator/Documents/Kaas/Venom_ncRNA_project/Usable_data/Genome_files/Crotalus_viridis_annotation_with_BPP_and_myotoxin_2024.12.18.gff\n",
    "\n",
    "and fix re-add the three_prime_utr and five_prime_utr sequences back.\n",
    "\n",
    "Next step: BCFtools_get_fasta_file_three_prime_utr_2024.12.20.sh (/home/administrator/Documents/Kaas/Venom_ncRNA_project/Scripts/BCFtools/BCFtools_get_fasta_file_three_prime_utr_2024.12.20.sh)\n",
    "Previous step: None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import needed packages\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Laod file path into memory\n",
    "gtf = '/home/administrator/Documents/Kaas/Venom_ncRNA_project/Usable_data/Genome_files/CroVir_rnd1.all.maker.final.homologIDs.updatedNov2019_with_myos_geneidmod_edited_with_BPP.gtf'\n",
    "\n",
    "# Load GTF into memory\n",
    "gtf_df = pd.read_csv(\n",
    "    gtf, sep = '\\t', comment = '#', header = None, names= [\n",
    "        'seqid', 'source', 'type2', 'start', 'end', 'score', 'strand', 'phase', 'attributes2'\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Laod the file path of the GFF into memory\n",
    "gff = '/home/administrator/Documents/Kaas/Venom_ncRNA_project/Usable_data/Genome_files/Crotalus_viridis_annotation_with_BPP_and_myotoxin_2024.12.18.gff'\n",
    "\n",
    "# Load the GFF into memory\n",
    "gff_df = pd.read_csv(\n",
    "    gff, sep='\\t', comment='#', header=None, names = [\n",
    "        'seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['seqid', 'source', 'start', 'end', 'score', 'strand', 'phase']\n"
     ]
    }
   ],
   "source": [
    "# Get common columns for the two data frames\n",
    "shared_cols = gff_df.columns.intersection(gtf_df.columns).tolist()\n",
    "print(shared_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>type2</th>\n",
       "      <th>attributes2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PE-reconstructed-10x-myo</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1</td>\n",
       "      <td>2076</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=myotoxin_model_1;geneID=myotoxin1</td>\n",
       "      <td>gene</td>\n",
       "      <td>gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PE-reconstructed-10x-myo</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1</td>\n",
       "      <td>2076</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=myotoxin_model_1;geneID=myotoxin1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PE-reconstructed-10x-myo</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>1</td>\n",
       "      <td>746</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=myotoxin_model_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PE-reconstructed-10x-myo</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>1640</td>\n",
       "      <td>1765</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=myotoxin_model_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PE-reconstructed-10x-myo</td>\n",
       "      <td>.</td>\n",
       "      <td>exon</td>\n",
       "      <td>1911</td>\n",
       "      <td>2076</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=myotoxin_model_1</td>\n",
       "      <td>exon</td>\n",
       "      <td>gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372379</th>\n",
       "      <td>scaffold-un945</td>\n",
       "      <td>maker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3236</td>\n",
       "      <td>3291</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>gene_id \"maker-scaffold-un945-augustus-gene-0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372380</th>\n",
       "      <td>scaffold-un1500</td>\n",
       "      <td>maker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>778</td>\n",
       "      <td>780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>gene_id \"augustus_masked-scaffold-un1500-proce...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372381</th>\n",
       "      <td>scaffold-un2620</td>\n",
       "      <td>maker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>307</td>\n",
       "      <td>837</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>gene_id \"maker-scaffold-un2620-augustus-gene-0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372382</th>\n",
       "      <td>scaffold-un3951</td>\n",
       "      <td>maker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1109</td>\n",
       "      <td>1281</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>gene_id \"maker-scaffold-un3951-augustus-gene-0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372383</th>\n",
       "      <td>scaffold-un3951</td>\n",
       "      <td>maker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30</td>\n",
       "      <td>63</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>gene_id \"maker-scaffold-un3951-augustus-gene-0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>372384 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           seqid source        type  start   end score strand  \\\n",
       "0       PE-reconstructed-10x-myo      .  transcript      1  2076     .      +   \n",
       "1       PE-reconstructed-10x-myo      .  transcript      1  2076     .      +   \n",
       "2       PE-reconstructed-10x-myo      .        exon      1   746     .      +   \n",
       "3       PE-reconstructed-10x-myo      .        exon   1640  1765     .      +   \n",
       "4       PE-reconstructed-10x-myo      .        exon   1911  2076     .      +   \n",
       "...                          ...    ...         ...    ...   ...   ...    ...   \n",
       "372379            scaffold-un945  maker         NaN   3236  3291     .      +   \n",
       "372380           scaffold-un1500  maker         NaN    778   780     .      -   \n",
       "372381           scaffold-un2620  maker         NaN    307   837     .      -   \n",
       "372382           scaffold-un3951  maker         NaN   1109  1281     .      -   \n",
       "372383           scaffold-un3951  maker         NaN     30    63     .      -   \n",
       "\n",
       "       phase                            attributes            type2  \\\n",
       "0          .  ID=myotoxin_model_1;geneID=myotoxin1             gene   \n",
       "1          .  ID=myotoxin_model_1;geneID=myotoxin1       transcript   \n",
       "2          .               Parent=myotoxin_model_1             exon   \n",
       "3          .               Parent=myotoxin_model_1             exon   \n",
       "4          .               Parent=myotoxin_model_1             exon   \n",
       "...      ...                                   ...              ...   \n",
       "372379     .                                   NaN  three_prime_utr   \n",
       "372380     .                                   NaN  three_prime_utr   \n",
       "372381     .                                   NaN  three_prime_utr   \n",
       "372382     .                                   NaN   five_prime_utr   \n",
       "372383     .                                   NaN  three_prime_utr   \n",
       "\n",
       "                                              attributes2  \n",
       "0       gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...  \n",
       "1       gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...  \n",
       "2       gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...  \n",
       "3       gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...  \n",
       "4       gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...  \n",
       "...                                                   ...  \n",
       "372379  gene_id \"maker-scaffold-un945-augustus-gene-0....  \n",
       "372380  gene_id \"augustus_masked-scaffold-un1500-proce...  \n",
       "372381  gene_id \"maker-scaffold-un2620-augustus-gene-0...  \n",
       "372382  gene_id \"maker-scaffold-un3951-augustus-gene-0...  \n",
       "372383  gene_id \"maker-scaffold-un3951-augustus-gene-0...  \n",
       "\n",
       "[372384 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join the data frames and filter out any removing everthing unneaded from the GFF section\n",
    "gxf_df = (pd.merge(\n",
    "    gff_df, gtf_df,\n",
    "    on = shared_cols,\n",
    "    how = 'outer'\n",
    ")\n",
    ")\n",
    "gxf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix the ranges for the three_prime_utr and five_prime_utr and get proper feature values into the type column so that 3' and 5' UTRs aren't missing anymore\n",
    "gxf_df2 = (\n",
    "    gxf_df\n",
    "        .query('type2 != \"gene\"') # Filter out instances of gene IDs in that row to reduce data complexity\n",
    "        # Copy they three_prime_utr and five_prime_utrs into the main type column\n",
    "        # .assign(\n",
    "        #     type = lambda x: x['type2'].where(x['type2'].isin(['three_prime_utr', 'five_prime_utr']), x['type'])\n",
    "        # )\n",
    "        # .drop(columns = ['type2', 'attributes2']) # Remove columns I don't need\n",
    "        .drop_duplicates() # Equivalent to distinct() in R\n",
    "        # Add a new column to clarify what genes are what\n",
    "        .assign(\n",
    "            gene_id = lambda x: x['attributes2'].str.extract(r'gene_id \"([^\"]+)\"')\n",
    "        )\n",
    ")\n",
    "gxf_df2\n",
    "\n",
    "# Create a dictionary to map gene_id to attributes where type2 is 'exon'\n",
    "exon_attributes = gxf_df2[gxf_df2['type2'] == 'exon'].set_index('gene_id')['attributes'].to_dict()\n",
    "\n",
    "# Replace missing values in 'attributes' for 'three_prime_utr' or 'five_prime_utr' rows when 'type2' is 'exon'\n",
    "gxf_df2['attributes'] = gxf_df2.apply(\n",
    "    lambda row: exon_attributes.get(row['gene_id'], row['attributes']) if pd.isna(row['attributes']) else row['attributes'],\n",
    "    axis=1\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function that takes a data frame and re-calculates ranges for the CDS\n",
    "def range_recalc(df1, df2):\n",
    "\n",
    "    # Iterate through rows, checking if the data ranges overlap\n",
    "    # df1 is the CDS df\n",
    "    for i, rows1 in df1.iterrows():\n",
    "\n",
    "        # df2 is the 3'UTR df\n",
    "        for j, rows2 in df2.iterrows():\n",
    "            \n",
    "            # Check if the row is a three_prime_utr row, and if so do this stuff:\n",
    "            if rows2['type2'] == 'three_prime_utr':\n",
    "\n",
    "                # Check if the sequence IDs are the same and only continue if true\n",
    "                if (\n",
    "                    rows1['seqid'] == rows2['seqid'] \n",
    "                    and rows1['type'] == 'exon' \n",
    "                    and rows1['gene_id'] == rows2['gene_id']\n",
    "                ):\n",
    "\n",
    "                    print(f\"Matching seqid found: {rows1['seqid']}\")\n",
    "\n",
    "                    # Check if what strand the sequence is on\n",
    "                    # If the strand is sense do that\n",
    "                    if rows1['strand'] == '+':\n",
    "                        # Check if the ranges overlap\n",
    "                        if rows1['start'] < rows2['start'] and rows1['end'] == rows2['end']:\n",
    "                            # Adjust the end point for row1 (the CDS)\n",
    "                            df1.at[i, 'end'] = rows2['start'] - 1\n",
    "                            print(f\"Adjusted end for row {i}: {df1.at[i, 'end']}\")\n",
    "                        else:\n",
    "                            print(f\"Ranges do not overlap for row {i}\")\n",
    "                            continue\n",
    "\n",
    "                    # If the strand is anti-sense do this\n",
    "                    elif rows1['strand'] == '-':\n",
    "                        # Check if the ranges overlap\n",
    "                        if rows1['start'] == rows2['start'] and rows1['end'] > rows2['end']:\n",
    "                            # Adjust the start point for row1 (the CDS)\n",
    "                            df1.at[i, 'start'] = rows2['end'] + 1 \n",
    "                            print(f\"Adjusted start for row {i}: {df1.at[i, 'start']}\")\n",
    "                        else:\n",
    "                            print(f\"Ranges do not overlap for row {i}\")\n",
    "                            continue \n",
    "                    else:\n",
    "                        continue\n",
    "                else:\n",
    "                    continue\n",
    "\n",
    "\n",
    "            # Check if the row is a five_prime_utr row, and if so do this stuff:        \n",
    "            elif rows2['type'] == 'five_prime_utr':\n",
    "\n",
    "                # Check if the sequence IDs are the sam and only continue if true\n",
    "                if (\n",
    "                    rows1['seqid'] == rows2['seqid'] \n",
    "                    and rows1['type'] == 'exon' \n",
    "                    and rows1['gene_id'] == rows2['gene_id']\n",
    "                ):\n",
    "\n",
    "                    print(f\"Matching seqid found: {rows1['seqid']}\")\n",
    "\n",
    "                    # Check if what strand the sequence is on\n",
    "                    # If the strand is sense do that\n",
    "                    if rows1['strand'] == '+':\n",
    "                        # Check if the ranges overlap\n",
    "                        if rows1['start'] == rows2['start'] and rows1['end'] > rows2['end']:\n",
    "                            # Adjust the start point for row1 (the CDS)\n",
    "                            df1.at[i, 'start'] = rows2['end'] + 1 \n",
    "                            print(f\"Adjusted start for row {i}: {df1.at[i, 'start']}\")\n",
    "                        else:\n",
    "                            print(f\"Ranges do not overlap for row {i}\")\n",
    "                            continue \n",
    "                    \n",
    "                    # If the strand is anti-sense do this\n",
    "                    elif rows1['strand'] == '-':\n",
    "                        # Check if the ranges overlap\n",
    "                        if rows1['start'] < rows2['start'] and rows1['end'] == rows2['end']:\n",
    "                            # Adjust the end point for rows1 (the CDS)\n",
    "                            df1.at[i, 'end'] == rows2['start'] - 1\n",
    "                            print(f\"Adjusted end for row {i}: {df1.at[i, 'end']}\")\n",
    "                        else:\n",
    "                            print(f\"Ranges do not overlap for row {i}\")\n",
    "                            continue\n",
    "                    else:\n",
    "                        continue\n",
    "                else:\n",
    "                    continue\n",
    "\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "    return df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate CDS from everthing else\n",
    "cds_df = (\n",
    "    gxf_df2\n",
    "        .query(\"type == 'exon'\")\n",
    "        .drop_duplicates()\n",
    "        .query(\"gene_id == 'myotoxin1'\") # Filter out other gene_ids as a test if the function works\n",
    ")\n",
    "\n",
    "# Seperate 3' UTRs from everthing else\n",
    "three_utr_df = gxf_df2.query(\"type2 == 'three_prime_utr'\").drop_duplicates().query(\"gene_id == 'myotoxin1'\") # Filter out other gene_ids as a test if the function works\n",
    "\n",
    "# Seperate 5' UTRs from everthing else\n",
    "five_utr_df = gxf_df2.query(\"type2 == 'five_prime_utr'\").drop_duplicates().query(\"gene_id == 'myotoxin1'\") # Filter out other gene_ids as a test if the function works\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matching seqid found: PE-reconstructed-10x-myo\n",
      "Ranges do not overlap for row 2\n",
      "Matching seqid found: PE-reconstructed-10x-myo\n",
      "Ranges do not overlap for row 3\n",
      "Matching seqid found: PE-reconstructed-10x-myo\n",
      "Adjusted end for row 4: 1922\n",
      "                      seqid source  type  start   end score strand phase  \\\n",
      "2  PE-reconstructed-10x-myo      .  exon      1   746     .      +     .   \n",
      "3  PE-reconstructed-10x-myo      .  exon   1640  1765     .      +     .   \n",
      "4  PE-reconstructed-10x-myo      .  exon   1911  1922     .      +     .   \n",
      "\n",
      "                attributes type2  \\\n",
      "2  Parent=myotoxin_model_1  exon   \n",
      "3  Parent=myotoxin_model_1  exon   \n",
      "4  Parent=myotoxin_model_1  exon   \n",
      "\n",
      "                                         attributes2    gene_id  \n",
      "2  gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...  myotoxin1  \n",
      "3  gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...  myotoxin1  \n",
      "4  gene_id \"myotoxin1\"; transcript_id \"myotoxin_m...  myotoxin1  \n"
     ]
    }
   ],
   "source": [
    "# Use the function to re-calculate the start or end for the CDS\n",
    "three_utr_recalc_df = range_recalc(cds_df, three_utr_df)\n",
    "three_utr_recalc_df\n",
    "print(three_utr_recalc_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now that I know it works, I can apply the function to the entire data frame and give it a second data frame that contains both 3' and 5' UTRs\n",
    "# Seperate three_prime_utr and five_prime_utrs\n",
    "utrs_df = (\n",
    "    gxf_df2\n",
    "        .query(\"type in ['three_prime_utr', 'five_prime_utr']\")\n",
    "        .drop_duplicates()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the function to the main data frame and the above data frame\n",
    "recscaled_cds_df = range_recalc(gxf_df2, utrs_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
