#/bin/bash

<<Step5
This script is how I ran miRanda for the 3'UTR, which is the third step of the pipeline. This script
runs miRanda with the C. viridis genomes and the mir.fasta generated by ShortStack.
The previous steps of this pipeline are: BEDtools_script_for_3UTR_from_gtf_2024-4-9.sh (/home/administrator/Documents/Kaas/Venom_ncRNA_project/Scripts/BEDtools/Genome_Chunking_for_miRanda/2024-3-28/BEDtools_script_for_3UTR_from_gtf_2024-4-9.sh)
                                         BEDtools_script_for_5UTR_from_gtf_2024-4-9.sh (/home/administrator/Documents/Kaas/Venom_ncRNA_project/Scripts/BEDtools/Genome_Chunking_for_miRanda/2024-3-28/BEDtools_script_for_5UTR_from_gtf_2024-4-9.sh)
                                         BEDtools_script_for_Genes_from_gtf_2024-4-9.sh (/home/administrator/Documents/Kaas/Venom_ncRNA_project/Scripts/BEDtools/Genome_Chunking_for_miRanda/2024-3-28/BEDtools_script_for_Genes_from_gtf_2024-4-9.sh)
The next step in the pipeline is: miRanda_output_formating_2024-4-9.sh (/home/administrator/Documents/Kaas/Venom_ncRNA_project/Scripts/miRanda/miranda_formating/miRanda_output_formating_2024-4-9.sh)
Step5

# Define variables:

    # mirna_file or the RNA sequence to be compared to genomes:
    mirna_file="/home/administrator/Documents/Kaas/Venom_ncRNA_project/Usable_data/miRanda_mirna_inputs_from_shortstack/2024-4-9_Run/Post_clean/mature_mir.fasta"

    # genomes or the genomic DNA/RNA sequence, alignment procedure uses complementarity and not sequence identity:
    genomes=(
        "/home/administrator/Documents/Kaas/Venom_ncRNA_project/Usable_data/Genome_files/Cvv_2017_genome_with_myos_3UTR.fasta"
        "/home/administrator/Documents/Kaas/Venom_ncRNA_project/Usable_data/Genome_files/Cvv_2017_genome_with_myos_5UTR.fasta"
        "/home/administrator/Documents/Kaas/Venom_ncRNA_project/Usable_data/Genome_files/Cvv_2017_genome_with_myos_exons.fasta"
    )

    # Name of location for the output file:
    output_directory="/home/administrator/Documents/Kaas/Venom_ncRNA_project/Results/miRanda/miRanda_2024-4-9/"

# Run the program:
    # Actually run the program:
    for genome in "${genomes[@]}"; do
        # Extract the genome name without the path for the directory
        genome_name=$(basename "$genome" .fasta)

        # Create new file name for the output
        output_file="${output_directory}/${genome_name}_miranda_miRNA_targets.out"

        # Run miRanda
        miranda "$mirna_file" "$genome" -out "$output_file"

        # Report to me that the output has been complete
        echo "miRanda run for $genome_name complete."
    done